warehouse/query-core/src/main/java/datawave/next/retrieval/DocumentIterator.java

```diff
@@ -0,0 +1,201 @@
+import org.apache.accumulo.core.data.Value;
+import org.apache.accumulo.core.iterators.IteratorEnvironment;
+import org.apache.accumulo.core.iterators.SortedKeyValueIterator;
+import org.apache.commons.jexl3.parser.ASTJexlScript;
+import org.apache.commons.jexl3.parser.ParseException;
+import org.apache.hadoop.io.Text;
+
+import com.google.common.base.Preconditions;
+
+import datawave.query.attributes.Attribute;
+import datawave.query.attributes.AttributeFactory;
+import datawave.query.attributes.Document;
+import datawave.query.attributes.DocumentKey;
+import datawave.query.data.parsers.EventKey;
+import datawave.query.function.JexlEvaluation;
+import datawave.query.function.serializer.KryoDocumentSerializer;
+import datawave.query.jexl.DatawaveJexlContext;
+import datawave.query.jexl.HitListArithmetic;
+import datawave.query.jexl.JexlASTHelper;
+import datawave.query.util.Tuple3;
+
+/**
+ * An iterator that retrieves documents from the shard table.
+ */
+public class DocumentIterator extends DocumentIteratorOptions implements SortedKeyValueIterator<Key,Value> {
+
+    private Key tk = null;
+    private Value tv = null;
+
+    private Range range = null;
+    private Collection<ByteSequence> columnFamilies = null;
+    private boolean inclusive = false;
+
+    private final KryoDocumentSerializer serializer = new KryoDocumentSerializer();
+    private final List<Entry<Key,Value>> results = new LinkedList<>();
+
+    public DocumentIterator() {}
+
+    public DocumentIterator(DocumentIterator other, IteratorEnvironment env) {
+        this.source = other.source.deepCopy(env);
+        super.deepCopy(other);
+    }
+
+    @Override
+    public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> options, IteratorEnvironment env) throws IOException {
+        this.source = source;
+        this.options = options;
+        this.env = env;
+        validateOptions(this.options);
+    }
+
+    @Override
+    public boolean hasTop() {
+        if (tk == null) {
+            if (!results.isEmpty()) {
+                Entry<Key,Value> entry = results.remove(0);
+                tk = entry.getKey();
+                tv = entry.getValue();
+            }
+        }
+        return tk != null;
+    }
+
+    @Override
+    public void next() throws IOException {
+        tk = null;
+        tv = null;
+    }
+
+    @Override
+    public void seek(Range range, Collection<ByteSequence> columnFamilies, boolean inclusive) throws IOException {
+        this.range = range;
+        this.columnFamilies = columnFamilies;
+        this.inclusive = inclusive;
+
+        checkForScanRebuild();
+
+        // aggregate document
+        AttributeFactory attributeFactory = new AttributeFactory(typeMetadata);
+        EventKey parser = new EventKey();
+
+        for (String candidate : candidates) {
+
+            Range candidateRange = rangeForCandidate(candidate);
+            source.seek(candidateRange, excludeCFs, false);
```

Comment:
calling seek will not only seek, but also aggregate all the documents and evaluate them without calling next. This seems to break the normal iterator pattern? Also could there be memory concerns here with adding n documents regardless of size into memory?

Reply:
Created https://github.com/NationalSecurityAgency/datawave/issues/2915